added option `google-readability-namespace-comments.AllowNoNamespaceComments`

thorsten-klein opened the pull request:
new option AllowNoNamespaceComments for `google-readability-namespace-comments.AllowNoNamespaceComments` is added.

Fixes #124264

When true, the check will allow that no namespace comment is present. If a namespace comment is added but it is not matching, the check will fail. Default is `false`

thorsten-klein committed:
added option AllowNoNamespaceComments for google-readability-namespace-comments

new option AllowOmittingNamespaceComments for
google-readability-namespace-comments is added.

When `true`, the check will accept if no namespace comment is present.
The check will only fail if the specified namespace comment is different
than expected. Default is `false`.

vbvictor commented on clang-tools-extra/docs/clang-tidy/checks/llvm/namespace-comment.rst:
```diff
@@ -39,3 +39,10 @@ Options
+.. option:: AllowNoNamespaceComments
+
+   When true, the check will allow that no namespace comment is present.
+   If a namespace comment is added but it is not matching, the check will fail. Default is `false`.
```
Please, wrap line to 80 characters.

vbvictor replied:
Nit: I am not a native speaker but I suggest there could be an easier to understand wording:
```suggestion
   If a namespace comment is present but does not match namespace definition, the check will fail. Default is `false`.
```

vbvictor commented:
Please, update Release Notes in clang-tools-extra/docs/ReleaseNotes.rst

vbvictor commented on clang-tools-extra/docs/clang-tidy/checks/llvm/namespace-comment.rst:
```diff
@@ -39,3 +39,11 @@ Options
+
+   When true, the check will allow that namespace comments are ommitted
+   entirely. The check only fails if a namespace comment is present but does
+   not match. Default is `false`.
```
Nit: I think it will make easier to understand what it tries to match. The same should be corrected in clang-tools-extra/docs/ReleaseNotes.rst if you wish to apply this fix.
```suggestion
   not match namespace definition. Default is `false`.
```

vbvictor commented:
> What do you think about the option's name? Is `AllowNoNamespaceComments` fine for you or shall I rename to `AllowOmittingNamespaceComments` (or any other suggestion)?

As for now I can not think of a better name, maybe others will suggest.

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -402,6 +402,13 @@ Changes in existing checks
   <clang-tidy/checks/readability/use-std-min-max>` check to use correct template
   type in ``std::min`` and ``std::max`` when operand is integer literal.
 
+- Improved :doc:`google-readability-namespace-comments
```
Please keep alphabetical order (by check name) in this list.

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -402,6 +402,13 @@ Changes in existing checks
 
+- Improved :doc:`google-readability-namespace-comments
+  <clang-tidy/checks/google/readability-namespace-comments>` that permits 
+  omitting namespace comments entirely. With this option enabled, the check
```
One sentence is enough. Please mention option name.

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/llvm/namespace-comment.rst:
```diff
@@ -39,3 +39,11 @@ Options
+
+.. option:: AllowNoNamespaceComments
+
+   When true, the check will allow that namespace comments are ommitted
```
```suggestion
   When `true`, the check will allow that namespace comments are omitted
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/llvm/namespace-comment.rst:
```diff
@@ -39,3 +39,11 @@ Options
+   When true, the check will allow that namespace comments are ommitted
+   entirely. The check only fails if a namespace comment is present but does
+   not match namespace definition. Default is `false`.
+
```
Excessive newline.

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -287,6 +287,11 @@ Changes in existing checks
 
+- Improved :doc:`google-readability-namespace-comments
+  <clang-tidy/checks/google/readability-namespace-comments>` check to
+  add an option ``AllowNoNamespaceComments``, that permits
```
```suggestion
  add an option `AllowNoNamespaceComments`, that permits omitting namespace
```

Please use single back-ticks for option names and values. Please fill with text as much as 80-characters limit allows.

thorsten-klein replied:
I made it consistent to already existing ReleaseNote entry. 
Ref: https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/docs/ReleaseNotes.rst?plain=1#L334

> - Improved :doc:`modernize-use-integer-sign-comparison
  <clang-tidy/checks/modernize/use-integer-sign-comparison>` check to
  add an option ``EnableQtSupport``, that makes C++17 ``q20::cmp_*`` alternative
  available for Qt-based applications.

Shall I really use single back-tick or keep it consistent with double back-ticks as already existing?

EugeneZelenko replied:
`EnableQtSupport` should use single back-ticks. Sorry, I missed this in relevant pull-request.

EugeneZelenko commented on clang-tools-extra/clang-tidy/readability/NamespaceCommentCheck.cpp:
```diff
@@ -184,6 +189,11 @@ void NamespaceCommentCheck::check(const MatchFinder::MatchResult &Result) {
                                  : ("namespace '" + *NamespaceNameAsWritten + "'");
 
+  // If no namespace comment is allowed
+  if(!hasComment && AllowNoNamespaceComments) {
```
```suggestion
  if (!hasComment && AllowNoNamespaceComments) {
```

thorsten-klein commented:
@EugeneZelenko What do you think about some additional option `EnforceNoNamespaceComments` in order to enforce that no namespace comment is present?
If you like the idea, I could create some additional PR for introducing this option.

HerrCai0907 commented:
I am not a native English speaker, just a personal feeling. I prefer `AllowOmittingNamespaceComments` since it is clearer then `AllowNoNamespaceComments`.

HerrCai0907 reviewed (commented):
Could you organize test. It looks like copy from other tests and have some noise which has nothing to do with the function of the current PR.

Also, please rebase to main since release note is cleared

vbvictor commented:
@thorsten-klein, gentle ping, do you still wish to work on this check?
21th LLVM release will be soon, If you wish to have your changes in the upcoming release please rebase on main and fix issues suggested by HerrCai0907.

thorsten-klein commented:
Thank you for your gentle ping! 

I have rebased the branch to master and renamed the option to `AllowOmittingNamespaceComments` as suggested.

Regarding the tests, I'm unsure which ones should be removed. After reviewing them, I believe all the tests are necessary to ensure that enabling the option does not disrupt any existing functionality, since all other tests are executed with the option `AllowOmittingNamespaceComments: false`.

vbvictor reviewed (commented):
I suppose "organizing tests" means:
- remove redundant file "comments-missing-c++17.cpp". Check only your feature - omitted namespace - OK, wrong name - Bad.
In your case, I think It's even better to have a second RUN command with custom suffix in the original test file. Look at [narrowing-conversions-ignoreconversionfromtypes-option.cpp](https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/test/clang-tidy/checkers/bugprone/narrowing-conversions-ignoreconversionfromtypes-option.cpp) for reference. This will avoid copy of original test files.
- make your new namespaces smaller - I guess 2 lines will be enough because default value is 1

thorsten-klein commented on clang-tools-extra/docs/clang-tidy/checks/llvm/namespace-comment.rst:
```diff
@@ -39,3 +39,11 @@ Options
+
+   When true, the check will accept if no namespace comment is present.
+   The check will only fail if a namespace comment is specified which is
+   different than expeced. Defaults to `false`.
```
👍🏻

vbvictor commented on clang-tools-extra/test/clang-tidy/checkers/google/readability-namespace-comments-missing-nested-namespaces.cpp:
```diff
@@ -0,0 +1,19 @@
+// RUN:   '-config={CheckOptions: { \
+// RUN:     google-readability-namespace-comments.AllowOmittingNamespaceComments: true, \
+// RUN:     google-readability-namespace-comments.ShortNamespaceLines: 0, \
+// RUN:   }}'
```
Sorry for being not very clear, for option-testfile we should test the option itself plus a small basic test of core functionality (that incorrect names gets flagged).
That what I meant by "omitted namespace - OK, wrong name - Bad"

Am I wrong, and we should duplicate everything for option-testfiles? @carlosgalvezp

carlosgalvezp replied:
I agree with your interpretation, we should not duplicate the main file into the option file, but we still need to test the behaviors that we promise. As per the docs, we promise 2 behaviors:

-   When `true`, the check will accept if no namespace comment is present.
-   The check will only fail if the specified namespace comment is different
   than expected.

So the tests should reflect those behaviors when the option is True.

I think the newly added test case is fine, can you please add it also to the other test file (readability-namespace-comments-missing.cpp) for consistency?

carlosgalvezp reviewed (approved):
LGTM!

vbvictor commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -187,6 +187,11 @@ Changes in existing checks
   <clang-tidy/checks/concurrency/mt-unsafe>` check by fixing a false positive
   where ``strerror`` was flagged as MT-unsafe.
 
+- Improved :doc:`google-readability-namespace-comments
```
Could we add also an entry for `llvm-readability-namespace-comments` with the same text, `google-readability-namespace-comments` is an alias for it.

thorsten-klein replied:
> Shall I run the tests for this llvm-namespace-comment check as well (e.g. a separate RUN)?

OK. This does not make sense. Then google and llvm checks would be mixed in one .cpp file. And additionally the `CHECK-MESSAGES` is not compatible.

vbvictor reviewed (approved):
LGTM

vbvictor commented:
@thorsten-klein, by [LLVM Developer Policy](https://llvm.org/docs/DeveloperPolicy.html#email-addresses) you should have public email for making a contribution.
Please turn off [Keep my email addresses private](https://github.com/settings/emails) setting in your account.

thorsten-klein commented:
My e-mail address is now public 👍

carlosgalvezp commented:
@vbvictor it seems this was merged before the CI checks finished running, next time please wait for that ;) Could you or @thorsten-klein fix the broken autoformatter?

thorsten-klein commented:
I have no access to a computer today. I can fix tomorrow, if it has not been fixed before.

vbvictor commented:
I think I've waited for all checks to pass..
I think the main problem is that when this PR was first published, there wasn't a "clang-format job" so It didn't run on any commits before this got merged into main.

vbvictor commented:
> I think the main problem is that when this PR was first published, there wasn't a "clang-format job" so It didn't run on any commits before this got merged into main.

Actually, there is a failed job if you hover over a red cross next to one of the commits, my bad.
Follow-up PR is here.

carlosgalvezp commented:
What I saw is that some checks don't run automatically, but must be manually triggered (I guess by a maintainer). I did click on the button to run them and immediately after you merged the patch, good syncing :)

I understand this might be done to save CI resources, but clang-format is a very lightweight job that should always run.

carlosgalvezp commented:
Thanks for fixing quickly!
